vng_api_common/middleware.py

```diff
@@ -96,6 +101,111 @@ def has_scopes(self, scopes: Union[Scope, None]) -> bool:
+
+        if not applicaties:
+            auth_data = self._request_auth()
+            applicaties = self._save_auth(auth_data)
+
+        return applicaties
+
+    def _request_auth(self) -> list:
+        client = AuthorizationsConfig.get_client()
+
+        response = client.list(
+            'applicatie',
+            query_params={'client_ids': self.client_id}
+        )
+        return underscoreize(response['results'])
+
+    def _get_auth(self):
+        return Applicatie.objects.filter(client_ids__contains=[self.client_id])
+
+    def _save_auth(self, auth_data):
+        applicaties = []
+
+        for applicatie_data in auth_data:
+            applicatie_serializer = ApplicatieSerializer(data=applicatie_data)
+            applicatie_serializer.is_valid()
+            applicaties.append(applicatie_serializer.save())
+
+        return applicaties
+
+    @property
+    def client_id(self) -> Union[str, None]:
+        if self.encoded is None:
+            return None
+
+        # TODO extract client_id from payload
+        #  when token generation is changed
+        try:
+            header = jwt.get_unverified_header(self.encoded)
+        except jwt.DecodeError:
+            logger.info("Invalid JWT encountered")
+            raise PermissionDenied(
+                _('JWT could not be decoded. Possibly you made a copy-paste mistake.'),
+                code='jwt-decode-error'
+            )
+
+        try:
+            jwt_secret = JWTSecret.objects.get(identifier=header['client_identifier'])
+        except JWTSecret.DoesNotExist:
+            raise PermissionDenied(
+                'Client identifier bestaat niet',
+                code='invalid-client-identifier'
+            )
+        except KeyError:
+            raise PermissionDenied(
+                'Client identifier is niet aanwezig in JWT',
+                code='missing-client-identifier'
+            )
+        else:
+            key = jwt_secret.secret
+
+        try:
+            jwt.decode(self.encoded, key, algorithms='HS256')
+        except jwt.InvalidSignatureError as exc:
+            logger.exception("Invalid signature - possible payload tampering?")
+            raise PermissionDenied(
+                'Client credentials zijn niet geldig',
+                code='invalid-jwt-signature'
+            )
+
+        return header['client_identifier']
+
+    def has_auth(self, scopes, zaaktype, vertrouwelijkheidaanduiding) -> bool:
+
+        if scopes is None:
+            return False
+
+        applicaties = self.applicaties
+        scopes_provided = set()
+
+        for applicatie in applicaties:
+            # allow everything
+            if applicatie.heeft_alle_autorisaties is True:
+                return True
+
+            # consider all scopes at all zaaktypes and vertrouwelijkheidaanduiding
+            for autorisatie in applicatie.autorisaties.all():
+                if autorisatie.component == AuthorizationsConfig.get_solo().component \
```

Comment:
not a fan of looking up the config for every iteration, this can be moved out of all the loops. You can then also filter the `applicatie.autorisaties.filter(component=config.component)`

Reply:
looking at the next line -> you can even filter the `autorisaties` down to `.filter(zaaktype=zaaktype)`

Reply:
fixed